File: src/dashboard/apigateway/apigateway/common/contexts/context.py
# -*- coding: utf-8 -*-
#
# TencentBlueKing is pleased to support the open source community by making
# 蓝鲸智云 - API 网关(BlueKing - APIGateway) available.
# Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
# Licensed under the MIT License (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
#     http://opensource.org/licenses/MIT
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
# either express or implied. See the License for the specific language governing permissions and
# limitations under the License.
#
# We undertake not to change the open source license (MIT license) applicable
# to the current version of the project delivered to anyone in the future.
#
from abc import ABCMeta
from typing import Text

from django.utils.functional import cached_property

from apigateway.common.factories import SchemaFactory
from apigateway.core.constants import ContextScopeTypeEnum, ContextTypeEnum
from apigateway.core.models import Context

_undefined = object()


class BaseContext(metaclass=ABCMeta):
    scope_type: Text = ""
    type: Text = ""

    @cached_property
    def schema(self):
        raise NotImplementedError()

    def save(self, scope_id, config):
        return Context.objects.update_or_create(
            scope_type=self.scope_type,
            scope_id=scope_id,
            type=self.type,
            defaults={
                "config": config,
                "schema": self.schema,
            },
        )

    def get_config(self, scope_id, default=_undefined):
        try:
            return Context.objects.get(
                scope_type=self.scope_type,
                scope_id=scope_id,
                type=self.type,
            ).config
        except Context.DoesNotExist:
            if default is _undefined:
                raise

            return default

    def filter_contexts(self, scope_ids=None):
        queryset = Context.objects.filter(
            scope_type=self.scope_type,
            type=self.type,
        )
        if scope_ids is not None:
            queryset = queryset.filter(scope_id__in=scope_ids)
        return queryset

    def filter_scope_id_config_map(self, scope_ids=None):
        return {context.scope_id: context.config for context in self.filter_contexts(scope_ids)}

    def delete(self, scope_ids):
        return Context.objects.filter(
            scope_type=self.scope_type,
            scope_id__in=scope_ids,
            type=self.type,
        ).delete()


class APIFeatureFlagContext(BaseContext):
    scope_type = ContextScopeTypeEnum.API.value
    type = ContextTypeEnum.API_FEATURE_FLAG.value

    @cached_property
    def schema(self):
        return SchemaFactory().get_context_api_feature_flag_schema()


class StageRateLimitContext(BaseContext):
    scope_type = ContextScopeTypeEnum.STAGE.value
    type = ContextTypeEnum.STAGE_RATE_LIMIT.value

    @cached_property
    def schema(self):
        return SchemaFactory().get_context_stage_rate_limit_schema()


class StageProxyHTTPContext(BaseContext):
    """Context data related with HTTP proxy in "stage" scope."""

    scope_type = ContextScopeTypeEnum.STAGE.value
    type = ContextTypeEnum.STAGE_PROXY_HTTP.value

    @cached_property
    def schema(self):
        return SchemaFactory().get_context_stage_proxy_http_schema()

    def contain_hosts(self, scope_id: int) -> bool:
        """Check if the config contains any valid host entries. The hosts were set
        to empty by default and require manual setup.

        :param scope_id: The ID of stage.
        :return: Whether the config contains valid host.
        """
        cfg = self.get_config(scope_id, default=None)
        if not cfg:
            return False
        return any(h.get("host") for h in cfg["upstreams"].get("hosts", []))


class ResourceAuthContext(BaseContext):
    scope_type = ContextScopeTypeEnum.RESOURCE.value
    type = ContextTypeEnum.RESOURCE_AUTH.value

    @cached_property
    def schema(self):
        return SchemaFactory().get_context_resource_bkauth_schema()


class APIAuthContext(BaseContext):
    scope_type = ContextScopeTypeEnum.API.value
    type = ContextTypeEnum.API_AUTH.value

    @cached_property
    def schema(self):
        return SchemaFactory().get_context_api_bkauth_schema()

File: src/mcp-proxy/pkg/util/context.go
/*
 * TencentBlueKing is pleased to support the open source community by making
 * 蓝鲸智云 - API 网关(BlueKing - APIGateway) available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 *     http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * We undertake not to change the open source license (MIT license) applicable
 * to the current version of the project delivered to anyone in the future.
 */

// Package util ...
package util

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"mcp_proxy/pkg/constant"
)

// SetBkUsername ...
func SetBkUsername(c *gin.Context, userName string) {
	c.Set(string(constant.BkUsername), userName)
	if c.Request != nil {
		c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.BkUsername, userName))
	}
}

// SetBkAppCode ...
func SetBkAppCode(c *gin.Context, appCode string) {
	c.Set(string(constant.BkAppCode), appCode)
	if c.Request != nil {
		c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.BkAppCode, appCode))
	}
}

// GetBkAppCode ...
func GetBkAppCode(c *gin.Context) string {
	appCode, ok := c.Get(string(constant.BkAppCode))
	if !ok {
		return ""
	}
	return appCode.(string)
}

// GetAppCode is an alias for GetBkAppCode
func GetAppCode(c *gin.Context) string {
	return GetBkAppCode(c)
}

// GetAppCodeFromContext gets app code from context
func GetAppCodeFromContext(ctx context.Context) string {
	if appCode, ok := ctx.Value(constant.BkAppCode).(string); ok {
		return appCode
	}
	return ""
}

// GetUsernameFromContext gets username from context
func GetUsernameFromContext(ctx context.Context) string {
	if username, ok := ctx.Value(constant.BkUsername).(string); ok {
		return username
	}
	return ""
}

// SetInnerJWTToken ...
func SetInnerJWTToken(c *gin.Context, jwtToken string) {
	c.Set(string(constant.BkGatewayInnerJWT), jwtToken)
	if c.Request != nil {
		c.Request = c.Request.WithContext(
			context.WithValue(c.Request.Context(), constant.BkGatewayInnerJWT, jwtToken))
	}
}

// SetMCPServerID ...
func SetMCPServerID(c *gin.Context, mcpServerID int) {
	c.Set(string(constant.MCPServerID), mcpServerID)
	if c.Request != nil {
		c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.MCPServerID, mcpServerID))
	}
}

// SetMCPServerName ...
func SetMCPServerName(c *gin.Context, mcpServerName string) {
	c.Set(string(constant.MCPServerName), mcpServerName)
	if c.Request != nil {
		c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.MCPServerName, mcpServerName))
	}
}

// GetMCPServerName ...
func GetMCPServerName(c *gin.Context) string {
	mcpServerName, ok := c.Get(string(constant.MCPServerName))
	if !ok {
		return ""
	}
	return mcpServerName.(string)
}

// GetMCPServerID ...
func GetMCPServerID(c *gin.Context) int {
	mcpServerID, ok := c.Get(string(constant.MCPServerID))
	if !ok {
		return 0
	}
	return mcpServerID.(int)
}

// GetMCPServerIDFromContext gets MCP server ID from context
func GetMCPServerIDFromContext(ctx context.Context) int {
	if mcpServerID, ok := ctx.Value(constant.MCPServerID).(int); ok {
		return mcpServerID
	}
	return 0
}

// SetGatewayID ...
func SetGatewayID(c *gin.Context, gatewayID int) {
	c.Set(string(constant.GatewayID), gatewayID)
	if c.Request != nil {
		c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.GatewayID, gatewayID))
	}
}

// GetGatewayID ...
func GetGatewayID(c *gin.Context) int {
	mcpServerID, ok := c.Get(string(constant.GatewayID))
	if !ok {
		return 0
	}
	return mcpServerID.(int)
}

// GetGatewayIDFromContext ...
func GetGatewayIDFromContext(ctx context.Context) int {
	gatewayID, ok := ctx.Value(constant.GatewayID).(int)
	if !ok {
		return 0
	}
	return gatewayID
}

// GetInnerJWTTokenFromContext ...
func GetInnerJWTTokenFromContext(ctx context.Context) string {
	jwtToken := ctx.Value(constant.BkGatewayInnerJWT)
	if innerJwt, ok := jwtToken.(string); ok {
		return innerJwt
	}
	return ""
}

// SetBkApiTimeout ...
func SetBkApiTimeout(c *gin.Context, timeout int) {
	c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.BkApiTimeout, timeout))
}

// GetBkApiTimeout returns the timeout duration for the BK API call
func GetBkApiTimeout(ctx context.Context) time.Duration {
	// Get the timeout value from the context
	timeout, ok := ctx.Value(constant.BkApiTimeout).(int)
	if !ok || timeout == 0 {
		// default timeout is 5 minute
		return 5 * time.Minute
	}
	return time.Duration(timeout) * time.Second
}

// SetBkApiAllowedHeaders ... 设置允许的请求头
func SetBkApiAllowedHeaders(c *gin.Context, allowedHeaders string) {
	allowedHeadersMap := make(map[string]string)
	for _, header := range strings.Split(allowedHeaders, ",") {
		header = strings.TrimSpace(header)
		if header == "" {
			continue
		}
		allowedHeadersMap[header] = c.Request.Header.Get(header)
	}
	// 默认添加 mcp-server 相关请求头
	allowedHeadersMap[constant.BkApiMCPServerIDKey] = fmt.Sprintf("%d", GetMCPServerID(c))
	allowedHeadersMap[constant.BkApiMCPServerNameKey] = GetMCPServerName(c)
	c.Request = c.Request.WithContext(context.WithValue(c.Request.Context(), constant.BkApiAllowedHeaders,
		allowedHeadersMap))
}

// GetBkApiAllowedHeaders ... 获取允许的请求头
func GetBkApiAllowedHeaders(ctx context.Context) map[string]string {
	// Get the timeout value from the context
	allowedHeaders, ok := ctx.Value(constant.BkApiAllowedHeaders).(map[string]string)
	if !ok {
		return map[string]string{}
	}
	return allowedHeaders
}

// JWTClaimsForLazySigning 用于延迟签发 JWT 的 claims 信息
type JWTClaimsForLazySigning struct {
	AppCode      string
	AppVerified  bool
	Username     string
	UserVerified bool
	Issuer       string
	Audience     []string
}

// SetJWTClaimsForLazySigning 保存 JWT claims 和私钥到 context，用于延迟签发
func SetJWTClaimsForLazySigning(c *gin.Context, claims *JWTClaimsForLazySigning, privateKey []byte) {
	c.Set(string(constant.BkGatewayJWTClaims), claims)
	c.Set(string(constant.BkGatewayPrivateKey), privateKey)
	if c.Request != nil {
		ctx := c.Request.Context()
		ctx = context.WithValue(ctx, constant.BkGatewayJWTClaims, claims)
		ctx = context.WithValue(ctx, constant.BkGatewayPrivateKey, privateKey)
		c.Request = c.Request.WithContext(ctx)
	}
}

// GetJWTClaimsFromContext 从 context 获取 JWT claims
func GetJWTClaimsFromContext(ctx context.Context) *JWTClaimsForLazySigning {
	if claims, ok := ctx.Value(constant.BkGatewayJWTClaims).(*JWTClaimsForLazySigning); ok {
		return claims
	}
	return nil
}

// GetPrivateKeyFromContext 从 context 获取私钥
func GetPrivateKeyFromContext(ctx context.Context) []byte {
	if key, ok := ctx.Value(constant.BkGatewayPrivateKey).([]byte); ok {
		return key
	}
	return nil
}
